src/main/java/com/puppycrawl/tools/checkstyle/xpath/AbstractNode.java

```diff
@@ -139,13 +139,31 @@ public AxisIterator iterateAxis(byte axisNumber, NodeTest nodeTest) {
     }
 
     /**
-     * Compares current object with specified for order.
+     * Compares two nodes by their appearance in document (document ordering).
+     * Result is negative if:
+     * <ul>
+     *     <li>Line number of the current object is lower than the
+     *     {@code nodeInfo} line number</li>
+     *     <li>Line numbers are same, but column number of the current object is lower than the
+     *     {@code nodeInfo} column number</li>
+     *     <li>Line number and column number are same, but text of the current node is
+     *     lexicographically lower than the {@code nodeInfo} text</li>
```

Comment:
Is lexicographically compare just to make this easier to code? Since we want appearance order, shouldn't the order in this case be parent/child/sibling relationship check? (IE: parent comes before child, first sibling comes before second sibling, etc..)

Reply:
This is good point.
It should be non rate case when parent node share the same line and column with first child. BUT it never happens for nodes with same depth in tree (siblings).
So we can count depth for nodes , if depth is the same throw exception, as it means that tree is damaged.

Reply:
> It should be non rate case when parent node share the same line and column with first child. BUT it never happens for nodes with same depth in tree (siblings).

I can't speak to rarity, but this statement is not true, especially that it never happens with node with same depth (siblings).

Example Code:
````
public class TestClass {
    void method() {
    }
}
````
Tree:
````
CLASS_DEF -> CLASS_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|  `--LITERAL_PUBLIC -> public [1:0]
|--LITERAL_CLASS -> class [1:7]
|--IDENT -> TestClass [1:13]
`--OBJBLOCK -> OBJBLOCK [1:23]
    |--LCURLY -> { [1:23]
    |--METHOD_DEF -> METHOD_DEF [2:4]
    |  |--MODIFIERS -> MODIFIERS [2:4]
    |  |--TYPE -> TYPE [2:4]
    |  |  `--LITERAL_VOID -> void [2:4]
    |  |--IDENT -> method [2:9]
    |  |--LPAREN -> ( [2:15]
    |  |--PARAMETERS -> PARAMETERS [2:16]
    |  |--RPAREN -> ) [2:16]
    |  `--SLIST -> { [2:18]
    |      `--RCURLY -> } [3:4]
    `--RCURLY -> } [4:0]
````

`OBJBLOCK` and `LCURLY` are parent/child and have the same line/column. Same with `TYPE` and `LITERAL_VOID`.
`MODIFIERS` and `TYPE` are siblings and have the same lime/column. Same with `PARAMETERS` and `RPAREN`.

I specifically remember this from another issue that if we have no line/column for a token, it looks at tokens around it and copies their line/column. So it is still a good possibility that 2 different tokens, regardless of parent/sibling, can use the same line/column.

Reply:
@rnveach good point, I will check it

Reply:
> 
```
    |--METHOD_DEF -> METHOD_DEF [2:4]
    |  |--MODIFIERS -> MODIFIERS [2:4]
    |  |--TYPE -> TYPE [2:4]
```

heh, ... ok lets throw no exceptions. If depth the same check position among siblings.

Reply:
we can take a look at other implementations of such compare method in other projects.

If this method can take as parameter any node is Tree ( so they might have as common node only Root node) we might need to make for each node list of indexes among siblings( for `TYPE [2:4]` it will be `3,1,1`, for `MODIFIERS [2:4]` it will be `3,1,0`) and compare such lists to make decision what is bigger (it should be `TYPE`).

Reply:
Current implementation of the `ElementNode` already has `indexAmongSiblings` field, so I can use it in `compareOrder` method,  what do you think?